lib/stores/S3Store.js

```diff
@@ -276,57 +276,81 @@ class S3Store extends DataStore {
-     *
-     * @param {Object}                metadata            upload metadata
-     * @param {http<IncomingMessage>} req                 incoming request
-     * @param {Number}                current_part_number number of the current part/chunk
-     * @return {Promise<Number>}                         which resolves with the current offset
-     * @memberof S3Store
-     */
-    _processUpload(metadata, req, current_part_number) {
-        return this._uploadPart(metadata, req, current_part_number)
-            .then(() => this.getOffset(metadata.file.id, true))
-            .then((current_offset) => {
-                if (parseInt(metadata.file.upload_length, 10) === current_offset.size) {
-                    return this._finishMultipartUpload(metadata, current_offset.parts)
-                        .then((location) => {
-                            log(`[${metadata.file.id}] finished uploading: ${location}`);
-
-                            this.emit(EVENTS.EVENT_UPLOAD_COMPLETE, {
-                                file: Object.assign({}, metadata.file, { location }),
-                            });
-
-                            this._clearCache(metadata.file.id);
-
-                            return current_offset.size;
-                        })
-                        .catch((err) => {
-                            throw err;
+   * uploads a stream to s3 using multiple parts
+   *
+   * @param {Object}                metadata            upload metadata
+   * @param {http<IncomingMessage>} req                 incoming request
+   * @param {Number}                currentPartNumber number of the current part/chunk
+   * @return {Promise<Number>}                         which resolves with the current offset
+   * @memberof S3Store
+   */
+    _processUpload(metadata, req, currentPartNumber) {
+        return new Promise((resolve, reject) => {
+            const splitter = new FileStreamSplitter(
+                {},
+                { maxChunkSize: this.part_size, directory: '.' }
+            );
+
+            const promises = [];
+
+            let pendingChunkFilepath = null;
+
+            stream.pipeline(req, splitter, (pipelineErr) => {
+                if (pipelineErr) {
+                    if (pendingChunkFilepath !== null) {
+                        fs.rm(pendingChunkFilepath, (fileRemoveErr) => {
+                            if (fileRemoveErr) {
+                                log(
+                                    `[${metadata.file.id}] failed to remove chunk ${pendingChunkFilepath}`
+                                );
+                            }
                         });
+                    }
                 }
 
-                return current_offset.size;
+                promises.push(pipelineErr ? Promise.reject(pipelineErr) : Promise.resolve());
+                resolve(promises);
+            });
+
+            splitter.on('chunkStarted', (filepath) => {
+                pendingChunkFilepath = filepath;
```

Comment:
pendingChunkFilepath always points to the file that the FileStreamSplitter current writes to, doesn't it? So it is never possible to have two chunkStarted events after another without one chunkFinished in between?

Reply:
Yes, but we only know the file path after chunk started